src/libraries/Common/src/Interop/Browser/Interop.Runtime.NativeAOT.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.JavaScript;
+
+internal static partial class Interop
+{
+    // WARNING: until https://github.com/dotnet/runtime/issues/37955 is fixed
+    // make sure that the native side always sets the out parameters
+    // otherwise out parameters could stay un-initialized, when the method is used in inlined context
+    internal static unsafe partial class Runtime
+    {
+#pragma warning disable SYSLIB1054
```

Comment:
Any particular reason to not use `LibraryImport`?

There is a sense in which these low-level functions shouldn't be using marshalling anyway, but at least there is nothing expensive here.

Reply:
> but at least there is nothing expensive here.

Actually, I take that back, there is UTF16 -> UTF8 marshalling going on.

It is quite inefficient and not really needed, since both C# and JS operate on 2-byte `char`s.

Reply:
Right, we should at least use UTF16, so that it pins the data instead of transcoding and copying it.

Ideally, upstream code would use the same approach, however, absent that we'll have to managed with `if (NativeAOT)` for now, i. e. something in the spirit of (@maraf):
```
; Warning: very wrong TS syntax
function mono_wasm_bind_js_function(function_name: any, ...)
    assert_bindings();
    if (!NativeAOT) {
    const function_name_root = mono_wasm_new_external_root<MonoString>(function_name),
        module_name_root = mono_wasm_new_external_root<MonoString>(module_name),
        resultRoot = mono_wasm_new_external_root<MonoObject>(result_address);

        const js_function_name = monoStringToString(function_name_root)!;
        const mark = startMeasure();
        const js_module_name = monoStringToString(module_name_root)!;
   } else {
        const js_function_name = utf16toString(function_name_root)!;
        const js_module_name = utf16toString(module_name_root)!;
   }
```

There is always a difficult act of touching less upstream code, but not so little that merging becomes conflict-free, yet broken functionality-wise. I think with these functions there is too much code (that we don't have expertise in, and that is fairly likely to change in the future) for them to be copied.

Reply:
Let's merge this here and see how it evolves. Once the dust settles, we could upstream some refactoring. For how I guess we don't really know how those "extension points" should look like in general and all the places we need them in.

Reply:
Right, not suggesting we do any kind of refactoring right now, hence the `if (NativeAOT)` path.

Reply:
Maraf just tried to make nice diff as you asked him earlier.